### PR TITLE
Feature/pa 38 user settings profile page

### PR DESCRIPTION
**Pull request by TondeMoon**
PA-38 User settings/Profile page

added endpoints (backend)
added pages, api call, test (frontend)
refactor history push action

## How it can be tested?
test for workWithMyData component

## PR Checklist

Please check if your PR fulfills the following requirements:
 - [X] I tested my code

### CODE REVIEW

File: backend/api/views.py [hunk 1]
```diff
@@ -14,3 +21,38 @@ class UserCreateAPIView(generics.CreateAPIView):
...
+    """
+    queryset = User.objects.all()
+    serializer_class = UserProfileSerializer
+    lookup_field = 'nickname'
+    permission_classes = (AllowAny,)
+
+    def get(self, request, nickname):
+        user_data = {
+            'nickname': nickname,
+            'email': f'{nickname}@gmail.com',
```
Comment (ivan-gerasin) [inline backend/api/views.py:38]: What about "show my email" flag? Can we already return actual user email?

File: frontend/src/App.scss [hunk 2]
```diff
@@ -4,3 +4,17 @@
...
 }
+
+.heading {
+  text-align: center;
+  font-size: 2rem;
+  color: white;
+  font-family:'Trebuchet MS', 'Lucida Sans Unicode', 'Lucida Grande', 'Lucida Sans', Arial, sans-serif
+}
+
+// @media (min-width: 1200px) {
```
Comment (ivan-gerasin) [inline frontend/src/App.scss]: Remove unused code

File: frontend/src/App.tsx [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
+import ProfilePage from '../src/components/personal/ProfilePage'
+import PreferencesPage from '../src/components/personal/PreferencesPage'
 import SignInSide from './components/authentication/SignIn'
+import {history} from './history'
 
 const App: React.FC = () => {
   
   return (
-    <Router>
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (ivan-gerasin) [inline frontend/src/App.tsx]: Why we disable ts check here?

File: frontend/src/actions/authActions.ts [hunk 4]
```diff
@@ -38,7 +39,12 @@ export const signUserUp = (payload: IUserDetails) => async (dispatch:Dispatch<fe
     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
     //@ts-ignore
     localStorage.setItem('token', result.token.access)
-    dispatch(setUserAction(result.user))
+       // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (ivan-gerasin) [inline frontend/src/actions/authActions.ts]: Please remove this comments. You can fix existing error, it should be easy.

File: frontend/src/actions/authActions.ts [hunk 5]
```diff
@@ -49,8 +55,14 @@ export const login = (payload: IUserDetails) => async (dispatch:Dispatch<fetchUs
 
   try {
     const result = await authApiClient.login(payload)
-    localStorage.setItem('token', result.access)
-    dispatch(setUserAction(result.user))
+    const a = result.access
```
Comment (ivan-gerasin) [inline frontend/src/actions/authActions.ts]: no singe-char variables please

File: frontend/src/actions/authActions.ts [hunk 6]
```diff
@@ -49,8 +55,14 @@ export const login = (payload: IUserDetails) => async (dispatch:Dispatch<fetchUs
 
   try {
     const result = await authApiClient.login(payload)
```
Comment (ivan-gerasin) [inline frontend/src/actions/authActions.ts]: You can specify return type of `authApiClient.login`

File: frontend/src/actions/authActions.ts [hunk 7]
```diff
@@ -49,8 +55,14 @@ export const login = (payload: IUserDetails) => async (dispatch:Dispatch<fetchUs
...
-    localStorage.setItem('token', result.access)
-    dispatch(setUserAction(result.user))
+    const a = result.access
+    localStorage.setItem('token', a)
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (ivan-gerasin) [inline frontend/src/actions/authActions.ts]: No disabling comments, unless they are really required.

File: frontend/src/actions/prefAndProfileActions.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
+/* eslint-disable no-magic-numbers */
```
Comment (ivan-gerasin) [inline frontend/src/actions/prefAndProfileActions.tsx]: ???

File: frontend/src/actions/prefAndProfileActions.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+export enum PrefActionType {
+  CHANGE_DATA = 'pref/CHANGE_DATA',
+  RETRIEVE_DATA= 'pref/RETRIEVE_DATA'
+}
+
+export interface IMyData {
+    nickname: string
+    email: string
+    id: string
```
Comment (ivan-gerasin) [inline frontend/src/actions/prefAndProfileActions.tsx]: Does id is really string? I think it probably a number. Please check, and fix if necessary.

File: frontend/src/actions/prefAndProfileActions.tsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+export type getUserAction = Action<PrefActionType.RETRIEVE_DATA, IUserPreference>
+export const postUserAction = actionCreator<PrefActionType.CHANGE_DATA, IUserPreference>(PrefActionType.CHANGE_DATA)
+
+export const getInqUserAction = actionCreator<PrefActionType.RETRIEVE_DATA, 
+IUserPreference>(PrefActionType.RETRIEVE_DATA)
+
+export const getMyData = () => async (dispatch:Dispatch<getUserAction>): Promise<void> => {
+
+  try {
+      const result = await workWithMyDataRequest.getData()
```
Comment (ivan-gerasin) [inline frontend/src/actions/prefAndProfileActions.tsx]: Please correct indent. I know, we should have perttier tool, I'll add it later. Until then, please watch for it.

File: frontend/src/api/HttpClient.tsx [hunk 11]
```diff
@@ -3,13 +3,23 @@ import {config} from '../config'
...
+  private async execute (url:string, method: string,  useCredentials: boolean, payload?: IUserDetails) {
+     const token = localStorage.getItem('token')
         const response = await fetch(`${baseURL}/${url}/`, {
           method: method,
           headers: {
             'Content-Type': 'application/json',
-            Accept: 'application/json',
+            Accept: 'application/json', 
+            Authorization: token&&!useCredentials? `Bearer ${token}` : '',
           },
```
Comment (ivan-gerasin) [inline frontend/src/api/HttpClient.tsx]: I would suggest next way:
```
const headers = {
  'Content-Type': 'application/json',
  Accept: 'application/json',
}
if (token&&!useCredentials) {
  headers['Authorization'] = `Bearer ${token}`
}
```
It makes reviewing request in dev-tools more clear. You will have 'Authorization' header _only_ if it has any content.

File: frontend/src/api/WorkWithMyData.tsx [hunk 12]
```diff
@@ -0,0 +1,27 @@
+/* eslint-disable @typescript-eslint/ban-ts-comment */
```
Comment (ivan-gerasin) [inline frontend/src/api/WorkWithMyData.tsx]: Can we remove it?

File: frontend/src/api/WorkWithMyData.tsx [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+
+  constructor(client: IHttpClient) {
+    this.client = client
+  }
+  
+    async getData(): Promise<IMyData> {
+      return await this.client.get('preferences', false) as IMyData
+      }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (ivan-gerasin) [inline frontend/src/api/WorkWithMyData.tsx]: I think it can be removed too

File: frontend/src/api/WorkWithMydata.test.ts [hunk 14]
```diff
@@ -0,0 +1,17 @@
+import {HttpClient} from './HttpClient'
+import {WorkWithMyData} from './WorkWithMyData'
+
+jest.mock('./HttpClient')
+const httpClient = new HttpClient()
+
+describe('use of WorkWithMyData client', () => {
+  beforeEach(() => {
+    HttpClient.mockClear()
```
Comment (ivan-gerasin) [inline frontend/src/api/WorkWithMydata.test.ts]: Where we mock any call? Do we really need it?

File: frontend/src/api/WorkWithMydata.test.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  })
+
+  it('WorkWithMyData should be able to call new() on HttpClient', () => {
```
Comment (ivan-gerasin) [inline frontend/src/api/WorkWithMydata.test.ts]: This test does not actually testing what is described.
You need to test that getData and saveData are make proper calls to HttpClient

File: frontend/src/api/authRequest.tsx [hunk 16]
```diff
@@ -1,19 +1,22 @@
...
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
 import {IUserAuthApiResponse, IUserDetails} from '../actions/authActions'
 import { HttpClient} from './HttpClient'
 
 
 export interface IAuthApiClient {
-    register: (payload: IUserDetails) => Promise<IUserAuthApiResponse>
-    login: (payload: IUserDetails) => Promise<IUserAuthApiResponse>
+    register: (payload: IUserDetails, useCredentials: true) => Promise<IUserAuthApiResponse>
```
Comment (ivan-gerasin) [inline frontend/src/api/authRequest.tsx]: You don't have useCredentials argument in actual register and login methods.
Also, you don't need it.

File: frontend/src/components/authentication/SignIn.tsx [hunk 17]
```diff
@@ -24,15 +23,13 @@ export default function SignInSide(): React.ReactElement {
   }, [setPassword])
 
   const dispatch = useDispatch()
-  const history = useHistory()
 
   const signIn = useCallback((e) => {
     e.preventDefault()
     // eslint-disable-next-line
```
Comment (ivan-gerasin) [inline frontend/src/components/authentication/SignIn.tsx]: Can we remove this comments?

File: frontend/src/components/personal/PreferencesPage.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+export default function PreferencesPage (): React.ReactElement {
+  const dispatch = useDispatch()
+  
+  useEffect(() => {
+        dispatch(getMyData())
+    }, [dispatch])  
+
+  // @ts-ignore
+  const umail = useSelector((state: RootState) => state.changeMyDataReducer.email)
```
Comment (ivan-gerasin) [inline frontend/src/components/personal/PreferencesPage.tsx]: `umail` => `email`

File: frontend/src/components/personal/PreferencesPage.tsx [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+import './PreferencesPage.scss'
+
+export default function PreferencesPage (): React.ReactElement {
+  const dispatch = useDispatch()
+  
+  useEffect(() => {
+        dispatch(getMyData())
+    }, [dispatch])  
+
+  // @ts-ignore
```
Comment (ivan-gerasin) [inline frontend/src/components/personal/PreferencesPage.tsx]: Please, work on this `// @ts-ignore` to remove it. You will need to specify return type of your reducer.

File: frontend/src/components/personal/PreferencesPage.tsx [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+  const [email, setEmail] = useState('')
+
+  const onChangeEmail = useCallback((e) => {
+    setEmail(e.target.value)
+  }, [setEmail])
+
+
+  const saveChanges = useCallback((e) => {
+    e.preventDefault()
+    // eslint-disable-next-line
```
Comment (ivan-gerasin) [inline frontend/src/components/personal/PreferencesPage.tsx]: I believe it can be fixed.

File: frontend/src/constants/headers.ts [hunk 21]
```diff
@@ -0,0 +1,13 @@
+
+export const token = localStorage.getItem('token')
```
Comment (ivan-gerasin) [inline frontend/src/constants/headers.ts]: Do we need this file?

File: frontend/src/history.ts [hunk 22]
```diff
@@ -0,0 +1,3 @@
+import { createBrowserHistory } from 'history'
+
+export const history = createBrowserHistory()
```
Comment (ivan-gerasin) [inline frontend/src/history.ts:3]: do we use it?

File: frontend/src/models/user.ts [hunk 23]
```diff
@@ -1,3 +1,9 @@
 export interface IUser {
   email: string
+}
+
+export interface IUserPreference{
+  nickname: string
+  email: string
+  id: string
```
Comment (ivan-gerasin) [inline frontend/src/models/user.ts]: is id a string?

File: frontend/src/reducers/changeMyDataReducer.tsx [hunk 24]
```diff
@@ -0,0 +1,37 @@
+import { PrefActionType, PrefActions } from '../actions/prefAndProfileActions'
+
+interface ChangeInitState {
+      // Here we use any becase type don't specified yet
+    // eslint-disable-next-line
+    nickname: any
```
Comment (ivan-gerasin) [inline frontend/src/reducers/changeMyDataReducer.tsx]: please specify types and get rid of `any`

**Review by ivan-gerasin — CHANGES_REQUESTED** [review]

File: frontend/src/api/HttpClient.tsx (hunk 11 above)
Comment (TondeMoon) [reply frontend/src/api/HttpClient.tsx]: TS is against of adding properties to object that were not there from the start - TS2339. I changed for your version, and its working - when it has empty Authorization in properties, but doesnt have Bearer null in it.

File: frontend/src/history.ts (hunk 22 above)
Comment (TondeMoon) [reply frontend/src/history.ts:3]: We use library `connected-react-router` for history.push action making through `React Router 5` after successful login operation, and its a demand of this lib that this operation should be in a separate file. I have `PA-56` task for refactoring this process, so lets leave at as it is by now.

File: frontend/src/models/user.ts (hunk 23 above)
Comment (TondeMoon) [reply frontend/src/models/user.ts]: We dont have any id by now from backend, but I corrected all to string type!

File: frontend/src/actions/authActions.ts (hunk 4 above)
Comment (TondeMoon) [reply frontend/src/actions/authActions.ts]: Its really easy, thanks!

File: frontend/src/api/authRequest.tsx (hunk 16 above)
Comment (TondeMoon) [reply frontend/src/api/authRequest.tsx]: Dont quite understand how its working, but I removed and everything still works.

File: frontend/src/actions/authActions.ts [hunk 25]
```diff
@@ -35,10 +39,11 @@ export const signUserUp = (payload: IUserDetails) => async (dispatch:Dispatch<fe
...
   try {
     const result = await authApiClient.register(payload)
-    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-    //@ts-ignore
     localStorage.setItem('token', result.token.access)
-    dispatch(setUserAction(result.user))
+    dispatch(setUserAction(result))
+     // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    //@ts-ignore
+    dispatch(push('/welcome'))
```
Comment (ivan-gerasin) [inline frontend/src/actions/authActions.ts]: You can remove this directives by doing next actions:
 - Import `CallHistoryMethodAction` from 'connected-react-router' in addition to 'push' that is already imported
 - Extend `Dispatch` type on line 37 from `dispatch:Dispatch<fetchUserAction>` to dispatch:Dispatch<fetchUserAction|CallHistoryMethodAction>

File: frontend/src/api/HttpClient.tsx [hunk 26]
```diff
@@ -1,16 +1,30 @@
...
+type AbstractApiData = any
+
+export interface IHttpClient {
+  post: <Payload, Response>(url:string, useCredentials: boolean,  payload: Payload) => Promise<Response>
+  get: <Response>(url:string, useCredentials: boolean) => Promise<Response>
+}
+
 export class HttpClient {
-    private async execute (url:string, payload: IUserDetails, method:string) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
Comment (ivan-gerasin) [inline frontend/src/api/HttpClient.tsx]: You can remove all this "no-explicit-any"

File: frontend/src/api/authRequest.tsx [hunk 27]
```diff
@@ -1,3 +1,6 @@
+/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
```
Comment (ivan-gerasin) [inline frontend/src/api/authRequest.tsx]: remove this

File: frontend/src/components/authentication/SignUp.tsx [hunk 28]
```diff
@@ -25,15 +23,13 @@ export default function SignUpSide(): React.ReactElement {
   }, [setPassword])
 
   const dispatch = useDispatch()
-  const history = useHistory()
 
   const signUp = useCallback((e) => {
     e.preventDefault()
     // eslint-disable-next-line
```
Comment (ivan-gerasin) [inline frontend/src/components/authentication/SignUp.tsx:29]: remove this

File: frontend/src/reducers/changeMyDataReducer.tsx [hunk 29]
```diff
@@ -0,0 +1,32 @@
+import { PrefActionType, PrefActions } from '../actions/prefAndProfileActions'
+
+interface ChangeInitState {
+    nickname: string
+      // eslint-disable-next-line
+    email: any
```
Comment (ivan-gerasin) [inline frontend/src/reducers/changeMyDataReducer.tsx]: Why `email` is any? 
`id` is highly likely a number

File: frontend/src/reducers/changeMyDataReducer.tsx [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+    id: any
+}
+
+export const initialState = {
+    nickname: '',
+    email: '',
+    id: ''
+}
+
+const changeMyDataReducer = ( state: ChangeInitState = initialState, action: PrefActions): ChangeInitState => {
```
Comment (ivan-gerasin) [inline frontend/src/reducers/changeMyDataReducer.tsx:15]: Please add tests for this reducer

**Review by ivan-gerasin — CHANGES_REQUESTED** [review]

**Review by ivan-gerasin — APPROVED** [review]

**Pull request merged**